Fix content/en/guide/v10/typescript.md (#718)
preactjs/preact-www: content/en/guide/v10/typescript.md

Preact ships TypeScript type definitions, which are used by the library itself! 

> When you use Preact in a TypeScript-aware editor (like VSCode), you can benefit from the added type information while writing regular JavaScript. If you want to add type information to your own applications, you can use [JSDoc annotations](https://fettblog.eu/typescript-jsdoc-superpowers/), or write TypeScript and transpile to regular JavaScript. This section will focus on the latter.

Author: This is a description of the VSCode features and methods of adding types.
I think it is unnecessary in Preact document.
Therefore, I removed this line.

Reviewer: A thought: may we could have this just be a short "tip" like this:

<img width="869" alt="Screen Shot 2020-12-07 at 10 31 46 AM" src="https://user-images.githubusercontent.com/105127/101370389-70395000-3877-11eb-9b93-3d79a2729f22.png">

the markup I used there was:

```md
> 💁 You can also use Preact's type definitions in plain JavaScript with [JSDoc annotations](https://fettblog.eu/typescript-jsdoc-superpowers/).
```

Rename your `.jsx` files to `.tsx` for TypeScript to correctly parse your JSX.

## Typing components
## Typing Components

> There are different ways to type components in Preact. Class components have generic type variables to ensure type safety. TypeScript sees a function as functional component as long as it returns JSX. There are multiple solutions to define props for functional components.

Author: > TypeScript sees a function as functional component as long as it returns JSX.

Unless a function is matched `FunctionComponent` or `FunctionalComponent`, it is not considered a functional components.
This is incorrect.

Author: I think this line is unnecessary.

Author: Since there is the same content at below, I think this line is unnecessary.

There are different ways to type components in Preact. Class components have generic type parameters to ensure type safety. TypeScript sees a function as functional component as long as it returns JSX. There are multiple solutions to define props for functional components.

> ### Function components

Author: They are "Functional Components" and "Class Components" in other page.
https://github.com/preactjs/preact-www/blob/922098858eba2769900cc909b18b7b27e5487a49/content/en/guide/v10/components.md#functional-components

Rename your `.jsx` files to `.tsx` for TypeScript to correctly parse your JSX.

## Typing Components

### Functional Components

The type corresponding to functional components is `FunctionComponent` type. This is type definition of `FunctionComponent` type:

```ts
interface FunctionComponent<P = {}> {
  (props: RenderableProps<P>, context?: any): VNode<any> | null;
  displayName?: string;
  defaultProps?: Partial<P>;
}
```

> 💁 `FunctionalComponent` type is alias of `FunctionComponent` type.

You can implement a functional component in TypeScript, as below.

```tsx
type Props = {
  name: string;
  age: number;
};

function MyComponent({ name, age }: Props) {
  return (
    <div>
      My name is {name}, I am {age} years old.
    </div>
  );

> }

Author: This example should be similar to the example of anonymous function.

Rename your `.jsx` files to `.tsx` for TypeScript to correctly parse your JSX.

## Typing Components

### Functional Components

The type corresponding to functional components is `FunctionComponent` type. This is type definition of `FunctionComponent` type:

```ts
interface FunctionComponent<P = {}> {
  (props: RenderableProps<P>, context?: any): VNode<any> | null;
  displayName?: string;
  defaultProps?: Partial<P>;
}
```

> 💁 FunctionalComponent type is alias of FunctionComponent type.

You can implement a functional component in TypeScript, as below.

```tsx
type Props = {
  name: string;
  age: number;
};

const MyComponent: FunctionComponent<Props> = function ({ name, age }) {
  return (
    <div>
      My name is {name}, I am {age} years old.
    </div>
  );
}
```

`children` is `ComponentChildren` type. `children` as `ComponentChildren` type is added to props in `RenderableProps` type:

```ts
type RenderableProps<P, RefType = any> = P & Readonly<Attributes & { children?: ComponentChildren; ref?: Ref<RefType> }>;

> ```

Author: Most React library use  `React.FC`.
In Preact, most library use `FunctionComponent`.
Using `FunctionComponent` method is general.
Therefore, This example is unnecessary.

```

You can set default props by setting a default value in the function signature.

```tsx
type GreetingProps = {
  name?: string; // name is optional!
}
`children` is `ComponentChildren` type. `children` as `ComponentChildren` type is added to props in `RenderableProps` type:

function Greeting({ name = "User" }: GreetingProps) {
  // name is at least "User"
  return <div>Hello {name}!</div>
}
```ts
type RenderableProps<P, RefType = any> = P & Readonly<Attributes & { children?: ComponentChildren; ref?: Ref<RefType> }>;
```

Preact also ships a `FunctionComponent` type to annotate anonymous functions. `FunctionComponent` also adds a type for `children`:
You do not need to add `children` as `ComponentChildren` type to `Props` type:

```tsx
import { h, FunctionComponent } from "preact";

const Card: FunctionComponent<{ title: string }> = ({ title, children }) => {
  return (
    <div class="card">
      <h1>{title}</h1>
      {children}
    </div>
  );
};
```

`children` is of type `ComponentChildren`. You can specify children on your own using this type:

```tsx
import { h, ComponentChildren } from "preact";

type ChildrenProps = {
type Props = {
  title: string;

>   children: ComponentChildren;

Author: Since `children` as `ComponentChildren` type is added to props in `RenderableProps` type, this line is unnecessary.

  componentDidUpdate?(
    previousProps: Readonly<P>,
    previousState: Readonly<S>,
    snapshot: any
  ): void;
  componentDidCatch?(error: any, errorInfo: any): void;
}

abstract class Component<P, S> {
  constructor(props?: P, context?: any);

  static displayName?: string;
  static defaultProps?: any;
  static contextType?: Context<any>;

  static getDerivedStateFromProps?(
    props: Readonly<object>,
    state: Readonly<object>
  ): object | null;
  static getDerivedStateFromError?(error: any): object | null;

  state: Readonly<S>;
  props: RenderableProps<P>;
  context: any;
  base?: Element | Text;

  setState<K extends keyof S>(
    state:
      | ((
          prevState: Readonly<S>,
          props: Readonly<P>
        ) => Pick<S, K> | Partial<S> | null)
      | (Pick<S, K> | Partial<S> | null),
    callback?: () => void
  ): void;

  forceUpdate(callback?: () => void): void;

  abstract render(
    props?: RenderableProps<P>,
    state?: Readonly<S>,
    context?: any
  ): ComponentChild;
}
```

You can implement a class component:

```tsx
import { h, Component } from 'preact';

// Types for props

> // in the case `children` is required.

Author: This line is unnecessary. I made mistake.

type ExpandableProps = {
  title: string;

>   children: ComponentChildren;

Author: This line is unnecessary. I made mistake.

}
```

You can set default props by setting a default value in the function signature.

```tsx
type GreetingProps = {
  name?: string; // name is optional!
}

function Greeting({ name = "User" }: GreetingProps) {
  // name is at least "User"
  return <div>Hello {name}!</div>
}

> ```

Author: This is explanation for Default parameters.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters
Therefore, This is not appropriate in TypeScript page of Preact document.

Reviewer 2: I think the section was great and we should keep it. People switching from prop types over to TS will naturally ask these questions as most linters treat prop-types and default props as the same. We should cater to those users too :+1:

  age: number;
};

> function MyComponent({ name, age }: Props) {

Author: Most React library use  `React.FC`.
In Preact, most library use `FunctionComponent`.
Using `FunctionComponent` method is general.
Therefore, This example is unnecessary.

}
```

> Preact ships a `FunctionComponent` type to annotate anonymous functions. `FunctionComponent` adds a type for `children` to `props`:

Author: This describes what a type for `children` is added to.
https://github.com/preactjs/preact/blob/f955cfcceb50d504faa7c05a95065512cf900e57/src/index.d.ts#L72-L90

}
```

> Preact also ships a `FunctionComponent` type to annotate anonymous functions. `FunctionComponent` also adds a type for `children`:

Author: `FunctionComponent` type corresponds all functional components, not only an anonymous function.
This sentence is confusing.

type RenderableProps<P, RefType = any> = P & Readonly<Attributes & { children?: ComponentChildren; ref?: Ref<RefType> }>;
```

> Since `children` in `RenderableProps` type is `children?: ComponentChildren`, `children` need to be specified when `children` is required:

Author: This clarifies the reason for specifying `children`.

### Class components

> Preact's `Component` class is a generic type. This type has two generic type variables which correspond Props and State. These generic type variables are empty object in default. You can define type and pass one as generic type variable, as below.

Reviewer: ```suggestion
Preact's `Component` class is a generic type. This type has two generic type variables which correspond Props and State. These generic type variables are empty objects by default, and can be defined as follows:
```

// ExpandableProps and ExpandableState are passed as generic type parameter.
class Expandable extends Component<ExpandableProps, ExpandableState> {
  constructor(props: ExpandableProps) {
    super(props);
    // this.state is an object with a boolean field `toggle`
    // due to ExpandableState
    // `this.state` is a object with `toggle` property which is boolean type by ExpandableState.
    this.state = {
      toggled: false
    };
  }
  // `this.props.title` is string due to ExpandableProps
  render() {
    return (
      <div class="expandable">
        <h2>
          // `this.props.title` is string type by ExpandableProps.
          {this.props.title}{" "}
          <button
            onClick={() => this.setState({ toggled: !this.state.toggled })}
          >
          <button onClick={() => this.setState({ toggled: !this.state.toggled })}>
            Toggle
          </button>
        </h2>
        // `this.props.children` is `ComponentChildren` type by default.
        <div hidden={this.state.toggled}>{this.props.children}</div>
      </div>
    );
  }
}
```

Class components include children by default, typed as `ComponentChildren`.

> ## Typing events

Author: I think it is necessary to describe about `TargetedEvent`, since `TargetedEvent` does not exist in React.

};

// ExpandableProps and ExpandableState are passed as generic type variables.
class Expandable extends Component<ExpandableProps, ExpandableState> {
  constructor(props: ExpandableProps) {
    super(props);
    // `this.state` is a object with `toggle` property which is boolean type by ExpandableState.
    this.state = {
      toggled: false
    };
  }
  render() {
    return (
      <div class="expandable">
        <h2>

>           // `this.props.title` is string type by ExpandableProps.

Reviewer: I need to check if our syntax highlighter can handle line comments in JSX

          {this.props.title}{" "}
          <button onClick={() => this.setState({ toggled: !this.state.toggled })}>
            Toggle
          </button>
        </h2>
        // `this.props.children` is `ComponentChildren` type by default.
        <div hidden={this.state.toggled}>{this.props.children}</div>
      </div>
    );
  }
}
```

## Typing events

> Preact handles standard DOM events. When [lib](https://www.typescriptlang.org/tsconfig#lib) of TypeScript compiler options includes `DOM`, you can use standard DOM events types.

Reviewer: ```suggestion
Preact uses standard DOM events. Include `"DOM"` in TypeScript's [lib](https://www.typescriptlang.org/tsconfig#lib) compiler option to enable standard DOM event types.
```

      </div>
    );
  }
}
```

Class components include children by default, typed as `ComponentChildren`.

## Typing events

Preact emits regular DOM events. As long as your TypeScript project includes the `dom` library (set it in `tsconfig.json`), you have access to all event types that are available in your current configuration.
Preact handles standard DOM events. When [lib](https://www.typescriptlang.org/tsconfig#lib) of TypeScript compiler options includes `DOM`, you can use standard DOM events types.

```tsx
export class Button extends Component {
  handleClick(event: MouseEvent) {
    event.preventDefault();
    if (event.target instanceof HTMLElement) {

>       alert(event.target.tagName); // Alerts BUTTON

Author: Another one uses `console.log()`.
It is inconsistent.

Reviewer: Good call!

## Typing events

> Preact emits regular DOM events. As long as your TypeScript project includes the `dom` library (set it in `tsconfig.json`), you have access to all event types that are available in your current configuration.

Author: > In Preact core, `onChange` is the standard [DOM change event]

There is "standard DOM event" in other page.
Therefore, I change to "standard DOM event".
Element emits event.

> Preact handles standard DOM events. As long as your TypeScript project includes the `dom` library (set it in `tsconfig.json`), you have access to all event types that are available in your current configuration.

Author: `lib` includes `DOM`.
`standard DOM events types` is more explicit.

}
```

> You can restrict event handlers by adding a type annotation for `this` to the function signature as the first argument. This argument will be erased after transpilation.

Author: This content is an explanation about [this parameters](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters).
Therefore, This is not appropriate in Preact document.

}
```

> You can specify a type of `EventTarget` by using [this parameters](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters), as below.

Reviewer: ```suggestion
You can provide a more specific type than `EventTarget` by using [this parameters](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters):
```

}
```

You can provide a more specific type than `EventTarget` by using [this parameters](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters):

```tsx
export class Button extends Component {
  // Define `this parameters`
  handleClick(this: HTMLButtonElement, event: MouseEvent) {
    event.preventDefault();
    console.log(this.tagName); // "BUTTON"

>   }

Author: I think this example should use `this`.
Since `this` is `HTMLButtonElement`, `instanceof` is unnecessary.
Another example uses `tagName`.

Reviewer 3: I agree, this is much more concise and showcases this binding much better 👍

## Typing references

The `createRef` function is also generic, and lets you bind references to element types. In this example, we ensure that the reference can only be bound to `HTMLAnchorElement`. Using `ref` with any other element lets TypeScript thrown an error:
`createRef()`is generic type. You can specify a reference type by passing `createRef()`'s generic type parameter:

```tsx
import { h, Component, createRef } from "preact";

class Foo extends Component {
  // `this.ref.current` is HTMLAnchorElement type.
  ref = createRef<HTMLAnchorElement>();

  componentDidMount() {

>     // current is of type HTMLAnchorElement

Author: `this.ref.current` is `div`.
This comment is confusing.

Reviewer 3: Good catch!

## Typing references

`createRef()`is [generic type](https://www.typescriptlang.org/docs/handbook/generics.html#generic-types). You can specify a reference type by passing it as `createRef()`'s generic type parameter:

```tsx
import { h, Component, createRef } from "preact";

class Foo extends Component {
  // `this.ref.current` is HTMLAnchorElement type.
  ref = createRef<HTMLAnchorElement>();

  render() {
    return <div ref={this.ref}>Foo</div>;
    //          ~~~
    //       💥 Error! `this.ref` can only be passed to Anchor Element's `ref` attribute.
  }
}
```

## Typing context

> `createContext` tries to infer as much as possible from the intial values you pass to:

Author: TypeScript infers, not `createContext`.

  }
  render() {
    return (
      <div class="expandable">
      <div>
        <h2>
          // `this.props.title` is string type by ExpandableProps.
          {this.props.title}{" "}
          <button onClick={() => this.setState({ toggled: !this.state.toggled })}>
            Toggle
          </button>
        </h2>
        // `this.props.children` is `ComponentChildren` type by default.
        // Since `this.props` is `RenderableProps<P>` type, `this.props.children` is `ComponentChildren` type by default.
        <div hidden={this.state.toggled}>{this.props.children}</div>
      </div>
    );
  }
}
```

## Typing Events
## Events

Preact uses standard DOM events. Include `"DOM"` in TypeScript's [lib](https://www.typescriptlang.org/tsconfig#lib) compiler option to enable standard DOM event types.

```tsx
import { h, Component } from 'preact';

export class Button extends Component {
  handleClick(event: MouseEvent) {

>     event.preventDefault();

Author: This line is unnecessary. Therefore, I removed it.

}
```

> Class components include children by default, typed as `ComponentChildren`.

Author: `children?` is default, not `children`.

```ts
type RenderableProps<P, RefType = any> = P & Readonly<Attributes & { children?: ComponentChildren; ref?: Ref<RefType> }>;
```

}
```

You can restrict event handlers by adding a type annotation for `this` to the function signature as the first argument. This argument will be erased after transpilation.
You can provide a more specific type than `EventTarget` by using [this parameters](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters):

```tsx
import { h, Component } from 'preact';

export class Button extends Component {
  // Adding the this argument restricts binding
  // Define `this parameters`
  handleClick(this: HTMLButtonElement, event: MouseEvent) {
    event.preventDefault();
    if (event.target instanceof HTMLElement) {
      console.log(event.target.localName); // "button"
    }
    console.log(this.tagName); // "BUTTON"
  }

  render() {
    return (
      <button onClick={this.handleClick}>{this.props.children}</button>

>     );

Author: This `(` and `)` are unnecessary. Therefore, I removed it.

Reviewer 2: This is the formatting that prettier applies by default and we have that on other pages of our docs

```tsx
import { h, Component } from 'preact';

export class Button extends Component {
  // Adding the this argument restricts binding
  // Define `this parameters`
  handleClick(this: HTMLButtonElement, event: MouseEvent) {
    event.preventDefault();
    if (event.target instanceof HTMLElement) {
      console.log(event.target.localName); // "button"
    }
    console.log(this.tagName); // "BUTTON"
  }

  render() {
    return (
      <button onClick={this.handleClick}>{this.props.children}</button>
    );
    return <button onClick={this.handleClick}>{this.props.children}</button>;
  }
}
```

You can provide a more specific type than `EventTarget` by using `TargetedEvent`. `TargetedEvent` type has two generic type parameters which corresponds `currentTarget`'s type and Event type:

```tsx
import { h, Component, JSX } from 'preact';

export class Button extends Component {
  handleClick({ currentTarget }: JSX.TargetedEvent<HTMLInputElement, Event> {
    console.log(currentTarget.tagName); // "BUTTON"
  }

  render() {
    return <button onClick={this.handleClick}>{this.props.children}</button>;
  }
}
```

## Typing references
## References

> The `createRef` function is also generic, and lets you bind references to element types. In this example, we ensure that the reference can only be bound to `HTMLAnchorElement`. Using `ref` with any other element lets TypeScript thrown an error:

Author: >  Using `ref` with any other element lets TypeScript thrown an error.

This is explanation about TypeScript `TypeError`.
I think it is unnecessary in Preact document.
Therefore, I removed this sentence.

`createRef()`is [generic type](https://www.typescriptlang.org/docs/handbook/generics.html#generic-types). You can specify a reference type by passing it as `createRef()`'s generic type parameter:

```tsx
import { h, Component, createRef } from "preact";

class Foo extends Component {
  // `this.ref.current` is HTMLAnchorElement type.
  ref = createRef<HTMLAnchorElement>();

  componentDidMount() {
    // current is of type HTMLAnchorElement
    console.log(this.ref.current);
  }

  render() {
    return <div ref={this.ref}>Foo</div>;
    //          ~~~

>     //       💥 Error! Ref only can be used for HTMLAnchorElement

Author: This content is repeating.

You can provide a more specific type than `EventTarget` by using [this parameters](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters):

```tsx
import { h, Component } from 'preact';

export class Button extends Component {
  // Define `this parameters`
  handleClick(this: HTMLButtonElement, event: MouseEvent) {

>     event.preventDefault();

Author: This line is unnecessary. Therefore, I removed it.

// }>
```

It also requires you to pass in all the properties you defined in the initial value:

```tsx
function App() {
  // This one errors 💥 as we haven't defined theme
  return (
    <AppContext.Provider
      value={{
//    ~~~~~ 
// 💥 Error: theme not defined
        lang: "de",
        authenticated: true
      }}
    >
    {}
      <ComponentThatUsesAppContext />
    </AppContext.Provider>
  );
}

> ```

Author: This is explanation about TypeScript `TypeError`.
I think it is unnecessary in Preact document.
Therefore, I removed this sentence.

If you don't want to specify all properties, you can either merge default values with overrides:

```tsx
const AppContext = createContext(appContextDefault);

function App() {
  return (
    <AppContext.Provider
      value={{
        lang: "de",
        ...appContextDefault
      }}
    >
      <ComponentThatUsesAppContext />
    </AppContext.Provider>
  );
}

> ```

Author: This is mere explanation about how to pass values to `props` using destructuring assignment.
I think it is not suitable in this page.
Therefore, I removed this sentence.

Or you work without default values and use bind the generic type variable to bind context to a certain type:

```tsx
type AppContextValues = {
  authenticated: boolean;
  lang: string;
  theme: string;
}

const AppContext = createContext<Partial<AppContextValues>>({});

function App() {
  return (
    <AppContext.Provider
      value={{
        lang: "de"
      }}
    >
      <ComponentThatUsesAppContext />
    </AppContext.Provider>
  );
```

> All values become optional, so you have to do null checks when using them.

Author: This is explanation about `Partial` type.
I think it is unnecessary in Preact document.
Therefore, I removed this sentence.

`useEffect` does extra checks so you only return cleanup functions.

```typescript
```ts
useEffect(() => {
  const handler = () => {
    document.title = window.innerWidth.toString();
  };
  window.addEventListener("resize", handler);

  // ✅  if you return something from the effect callback

>   // it HAS to be a function without arguments

Author: It seems this line is mere explanation about `useEffect`. 
No one passes an argument since this function does not accept arguments.
Therefore, this lines is unnecessary.

> `useEffect` does extra checks so you only return cleanup functions.

Author: It seems this line is mere explanation about `useEffect`.  Since Hooks page is suitable, this line is unnecessary.
